### DISCUSSION TITLE
Lesson 9 : Raffle Verification Error

### DISCUSSION THREAD

**Discussion by jojomensah89** [Cyfrin/foundry-full-course-cu#445] [Q&A] [answered]
After the contract is deployed to Sepolia, a request I s submitted for verification . I get this error. Can anyone help me out?
```
Start verification for (2) contracts
Start verifying contract `0xd693bd3890ba3751456a57fdc6b3209a4c588baf` deployed on sepolia

Submitting verification for [lib/foundry-devops/src/DevOpsTools.sol:DevOpsTools] "0xd693bd3890BA3751456A57fDC6B3209a4c588baF".

Submitting verification for [lib/foundry-devops/src/DevOpsTools.sol:DevOpsTools] "0xd693bd3890BA3751456A57fDC6B3209a4c588baF".

Submitting verification for [lib/foundry-devops/src/DevOpsTools.sol:DevOpsTools] "0xd693bd3890BA3751456A57fDC6B3209a4c588baF".
Submitted contract for verification:
        Response: `OK`
        GUID: `dszahgeggdzyccviirq5npyrsufwzehqjtehcp54zfnfc3lgbi`
        URL:
        https://sepolia.etherscan.io/address/0xd693bd3890ba3751456a57fdc6b3209a4c588baf
Contract verification status:
Response: `NOTOK`
Details: `Pending in queue`
Contract verification status:
Response: `NOTOK`
Details: `Fail - Unable to verify. Solidity Compilation Error: Source "lib/foundry-devops/lib/forge-std/src/Vm.sol" not found: File not found. Searched the following locations: "".`
Contract failed to verify.
make: *** [Makefile:43: deploy] Error 1
```

**Comment by shikhar229169** [comment]
Run the following command:
```bash
foundryup
```

And try again.

**Reply by jojomensah89** [reply]
Thanks for the reply, I have already tried it and it didn't work for me

**Reply by wuleiLonghe** [reply]
I'm getting same message,  Help!

**Reply by mikekochh** [reply]
Also running into the same issue

**Reply by mibiot** [reply]
+1 to the club) have got the same error

**Reply by mibiot** [reply]
to be precise - I changed the way to Vm.sol in DevOpsTools.sol to "lib/foundry-devops/lib/forge-std/src/Vm.sol"; This solves the error locally in my VSC:

<img width="1113" alt="Screenshot 2023-08-03 at 13 34 15" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/140085328/51611214-d422-4824-ac54-2b9d4c25d264">

 but when trying to update the contract's source code (to verify deployed contract manually) the issue is still there:
<img width="1173" alt="Screenshot 2023-08-03 at 13 34 38" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/140085328/efc52d71-069b-40de-b980-4cd67c68a2b7">

**Reply by mibiot** [reply]
I suppose cloning Patrick's repo will solve the problem, but since I have made so many comments during the course, and already have written some tests myself I would definitely prefer to leave my own version, somehow solving this problem. 
Much respect to those who can help.

**Reply by mibiot** [reply]
***UPDATE: - I changed both paths in DevOpsTools.sol to:

<img width="677" alt="Screenshot 2023-08-03 at 13 49 53" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/140085328/24ee63b5-6145-4946-a1f9-92c7736ed482">

and deployed with auto-verification again and it worked! Both contracts have been successfully deployed.

Hope this could help to those who will face this issue. Best

**Comment by 0xusmanf** [comment]
Can any of you provide GitHub repo of your project so I can recreate this error on my end?

**Reply by mikekochh** [reply]
https://github.com/mikekochh/Smart-Contract-Raffle

**Comment by Niferu** [comment]
Hello guys,

Do you have below remapping in your `foundry.toml` file?

`remappings = ["foundry-devops/=lib/foundry-devops"]`

My whole `foundry.toml` file below:
```toml
[profile.default]
src = "src"
out = "out"
libs = ["lib"]
remappings = ['@chainlink/contracts/=lib/chainlink-brownie-contracts/contracts/', '@solmate=lib/solmate/src/', "foundry-devops/=lib/foundry-devops"]
ffi = false

[etherscan]
mainnet = {key = "${ETHERSCAN_API_KEY}"}
sepolia = {key = "${ETHERSCAN_API_KEY}"}

[rpc_endpoints]
sepolia = "${SEPOLIA_RPC_URL}"
```

I have tested it now and including "foundry-devops/=lib/foundry-devops" solved issue for me and everything has been verified without any problems. (I have copied course repo code)

**Reply by mikekochh** [reply]
I added the "foundry-devops/=lib/foundry-devops" to the remappings and am still getting the same error. It is strange because I am able to deploy the moodNft smart contract and verify on etherscan fine, but when I try to run the script to mint an NFT, this is when it is unable to verify the contract. 

By the way, all of this works fine on anvil chain, it is only when I try to deploy on Sepolia where I run into issues.

**Reply by Niferu** [reply]
Not sure then what causing you this error as I cloned course repository, changed remappings only and it worked for me on Sepolia without any issues.

And by the way if you are calling mint function you are not deploying anything, so what do you want to be verified? Only deployed contracts can be verified and not its functions solo.

**Comment by theirrationalone** [comment]
# You Spammed the Sepolia testnet.

Please Follow in Order...
Do this first... (Don't do these reinstallations if you're sure with your prior installations).

<br>

```terminal
foundryup
```

<br>

1. delete: "chainlink-brownie-contract" Folder and "FoundryDevOps" Folder from lib.
2. Reinstall them,...

<br>

```terminal
forge install smartcontractkit/chainlink-brownie-contracts --no-commit
```

<br>

```terminal
sudo apt-get install jq
```

<br>

```terminal
forge install cyfrin/foundry-devops --no-commit
```

<br>

packages reinstallation was necessary because i don't know what you did(might messed up with) during installation.

<br>

# **CRUCIAL (Mandatory to do) 👇**

Now, Please Open metamak -> settings -> Advanced -> clear activity tab data (Button) to clear "Clear activity and nonce data".

<br>

After all of these 👆sanity checks rectifications. You need to check whether Alchemy's mempool has any pending requests or not... if you have any then you will have to wait or have to pay more gas to resolve or to complete them(Don't know how to do that, watch again Patrick's Video).

<br>

Now You're Good to go. Verify again.... Best of Luck!🍀

<br>
<br>

If Verification failed again after this 👆 mitigation too, then you will have to check your transaction's blocks, Check how many blocks are yet to be mined/passed.... (Your verification transaction will only be affected after blocks preceding get mined, In this case, Keeping Wait & Patience is The Only Mitigation).

Hope It Helps🤞🤞

Thank You 🙏🙏🙏🙏

**Reply by TreapGoGo** [reply, marked as answer]
I express my sincere appreciation for this answer! ❤️ 
I have encountered the same issue, but reinstalling Foundry and submodules really helped. 
By the way, I'd like to point out that the installation command you provided is not perfect, because running 
```bash
forge install smartcontractkit/chainlink-brownie-contracts --no-commit
```
 cannot install `chainlink-brownie-contracts` library completely. Some files (e.g. `VRFCoordinatorV2Mock.sol`) are just missing, but I don't know why.

Perfect installation command should be like this:
```bash
forge install smartcontractkit/chainlink-brownie-contracts@0.6.1 --no-commit
```

After the reinstallation, all my contracts got verified successfully. Thanks again for your help.

**Reply by theirrationalone** [reply]
Ohh my bad 😅, Actually i omitted to mention the version it too. But after all thanks ❤ for pointing that out.
